src/Modules/UngradedModules.jl

```diff
@@ -807,16 +908,211 @@ function base_ring(M::SubModuleOfFreeModule)
+    catch err
+      error("Submodule has no default ordering")
+    end
+  end
+  return M.default_ordering
+end
+
+@doc Markdown.doc"""
+    set_default_ordering!(M::SubModuleOfFreeModule, ord::ModuleOrdering)
+
+Set the default ordering in `M` to `ord`.
+"""
+function set_default_ordering!(M::SubModuleOfFreeModule, ord::ModuleOrdering)
+  M.default_ordering = ord
+end
+
 @doc Markdown.doc"""
     std_basis(submod::SubModuleOfFreeModule)
 
-Compute a standard basis of `submod`.
+Compute a standard basis of `submod` with respect to its default odering.
 """
 function std_basis(submod::SubModuleOfFreeModule)
-  if !isdefined(submod, :std_basis)
-    submod.std_basis = groebner_basis(submod.gens)
+  return std_basis(submod, default_ordering(submod))
+end
+
+@doc Markdown.doc"""
+    groebner_basis(submod::SubModuleOfFreeModule)
+
+Compute a Gröbner basis of `submod` with respect to its default odering.
+For this, the default ordering must be global.
+"""
+function groebner_basis(submod::SubModuleOfFreeModule)
+  return groebner_basis(submod, default_ordering(submod))
+end
+
+function std_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+  if !haskey(submod.groebner_basis, ordering)
+    gb = compute_std_basis(submod, ordering)
+    submod.groebner_basis[ordering] = gb
   end
-  return submod.std_basis
+  return submod.groebner_basis[ordering]
+end
+
+@doc Markdown.doc"""
+    groebner_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+
+Compute a Gröbner of `submod` with respect to the given `ordering`.
+The ordering must be global.
+"""
+function groebner_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+  @assert is_global(ordering)
+  return std_basis(submod, ordering)
+end
+
+@doc Markdown.doc"""
+    reduced_groebner_basis(submod::SubModuleOfFreeModule)
+
+Compute a reduced Gröbner basis. The return type is `ModuleGens`.
+"""
+function reduced_groebner_basis(submod::SubModuleOfFreeModule)
+  return reduced_groebner_basis(submod, default_ordering(submod))
+end
+
+@doc Markdown.doc"""
+    reduced_groebner_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+
+Compute a reduced Gröbner basis with respect to the given `ordering`. The return type is `ModuleGens`.
+"""
+function reduced_groebner_basis(submod::SubModuleOfFreeModule, ordering::ModuleOrdering)
+  @assert is_global(ordering)
+  
+  if !haskey(submod.groebner_basis, ordering)
+    gb = compute_std_basis(submod, ordering, true)
+    submod.groebner_basis[ordering] = gb
+    return gb
+  else
+    if !(submod.groebner_basis[ordering].is_reduced)
+      red_gb = get_attribute(submod.groebner_basis[ordering], :reduced_groebner_basis)
+      if red_gb === nothing
+        red_gb = compute_std_basis(submod, ordering, true)
+        set_attribute!(submod.groebner_basis[ordering], :reduced_groebner_basis => red_gb)
+      else
+        return red_gb
+      end
+    else
+      return submod.groebner_basis[ordering]
+    end
+  end
```

Comment:
All in all, perhaps do this (modulo testing that it actually is syntactically valid etc.):
```suggestion
  gb = get!(submod.groebner_basis, ordering) do
    return compute_std_basis(submod, ordering, true)
  end::ModuleGens
  gb.is_reduced && return gb
  return get_attribute!(gb, :reduced_groebner_basis) do
    return compute_std_basis(submod, ordering, true)
  end::ModuleGens
```

Reply:
Thanks, I didn't know this.